### PR TITLE
Added >>define command and added help text

### PR DESCRIPTION
**Pull request by johnmedina999**
Finally finished the `>>define` command. It is fully operational as of this moment, but can be improved on if it is necessary to do so.

### CODE REVIEW

**Review by abraker95 — CHANGES_REQUESTED** [review]
Remove the following files:
- VERSION
- runSigma.bat

Remove any inactive/commented code

**Comment by johnmedina999** [comment]
Added all requested changes.

File: sigma/plugins/utility/other/define.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        cmd.log("The user did not input a real word. Either that or the Merriam-Webster HTML files were changed, and the whole command needs to be reworked.")
+
+    rawDefArr = rawDef.split("<!--JC: actual definitions-->")
+
+    try:
+        soupObj = BeautifulSoup(rawDefArr[defNum], "html.parser")
+    except:
+        embed = discord.Embed(title=":exclamation: An error has occured. This number definition does not exist; please try again with a different number.", color=0xFF0000)
+        await message.channel.send(None, embed=embed)
+        cmd.log("This error should never trigger, but if it does, something is wrong.")
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: You can remove the cmd.log here. It tells us nothing more than what we already know (the discord message posting that the definition isn't found). cmd.log is used for reporting errors without printing how badly you goofed up in the discord chat.

File: sigma/plugins/utility/other/define.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    async with aiohttp.ClientSession() as session:
+        async with session.get(defURL) as resp:
+            rawDef = await resp.text()
+    
+    try:
+        rawDef = rawDef[rawDef.index("<!--JC: actual definitions-->")-10:]
+    except:
+        embed = discord.Embed(title=":unamused: An error has occured. Silly you! Enter a real word next time!", color=0xFF8EE6)
+        await message.channel.send(None, embed=embed)
+        cmd.log("The user did not input a real word. Either that or the Merriam-Webster HTML files were changed, and the whole command needs to be reworked.")
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: You can remove the cmd.log here. cmd.log is used for reporting errors without printing how badly you goofed up in the discord chat.

File: sigma/plugins/utility/other/define.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        except ValueError:
+            pass
+
+        if not defLast:
+            for i in args:
+                defURL += i
+                defURL += "%20"
+        else:
+            argsCopy = args.copy()
+            del(argsCopy[-1])
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: what is the del for?

File: sigma/plugins/utility/other/define.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    defNum = 1
+
+    defLast = False
+    
+    hasPrinted = False
+
+    defURL = r"https://www.merriam-webster.com/dictionary/"
+
+
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: Why leave so much whitespace here?

File: sigma/plugins/utility/other/define.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    if not args:
+        await message.channel.send(cmd.help())
+        return
+    else:
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: return ends the command. I recommend you remove the else and shift everything back by one indent. I try not to go into the right as much as possible to avoid spaghetti code

File: sigma/plugins/utility/other/define.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        try:
+            defNum = int(args[-1])
+            defLast = True
+        except ValueError:
+            pass
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: What does this part do?

File: sigma/plugins/utility/other/define.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        soupObj = BeautifulSoup(rawDefArr[defNum], "html.parser")
+    except:
+        embed = discord.Embed(title=":exclamation: An error has occured. This number definition does not exist; please try again with a different number.", color=0xFF0000)
+        await message.channel.send(None, embed=embed)
+        cmd.log("This error should never trigger, but if it does, something is wrong.")
+
+    for tag in soupObj.find_all(class_="intro-colon"):
+        definition += tag.text
+        if(len(tag.next_sibling.strip()) == 0):
+            if not hasPrinted:
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: `if hasPrinted: continue`
and shift the rest to the left by one indent

**Review by abraker95 — CHANGES_REQUESTED** [review]
There are some formatting issues with the code.

File: sigma/plugins/utility/other/define.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+                    definition += " : "
+                hasPrinted = True
+        else:
+            definition += tag.next_sibling
+            definition += " "
+
+    definition.strip()
+    definition.replace(":"," \n— ")
+
+    if not definition:
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: I am kinda concerned with how this done. If the website changes and "sx-link sc" is renamed to something else, as an example, then it will always print "That definition doesn't exist, silly". I need the code to differentiate between when the definition is not found and when the web code we are trying to scrape has changed, causing our code not to work on it. That's when I want the cmd.log to print to console an error that needs the bot dev's attention.

Also when `if not definition` is true, the previous lines should throw an exception since you are trying to access something that is not a valid object. This statement always has to be false or else it would never get there in the first place.

File: sigma/plugins/utility/other/define.py (hunk 6 above)
Comment (johnmedina999) [reply sigma/plugins/utility/other/define.py]: This part detects whether or not the user input a number as the last element in `args`. It tries to convert it to a string; if the user did in fact use a number, this will return `ValueError` which is ignored with `pass`.

File: sigma/plugins/utility/other/define.py (hunk 3 above)
Comment (johnmedina999) [reply sigma/plugins/utility/other/define.py]: `del()` removes the last element in `args`; this bit of code only runs if the user inputted a number as the last element. This avoids having the number be in the Merriam-Webster URL.

File: sigma/plugins/utility/other/define.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        try:
+            defNum = int(args[-1])
+            defLast = True
+        except ValueError:
+            pass
+
+        if not defLast:
+            for i in args:
+                defURL += i
+                defURL += "%20"
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: This can just be `defURL = r"https://www.merriam-webster.com/dictionary/" + args[0]` and avoid the this entire if/for statement chain. Are there any words that have spaces?

**Comment by johnmedina999** [comment]
Applied all changes except the ones I commented on.

File: sigma/plugins/utility/other/define.py (hunk 8 above)
Comment (johnmedina999) [reply sigma/plugins/utility/other/define.py]: If `That definition doesn't exist, silly` keeps popping up that is a sign that the command requires maintenance, which would be no different than a `cmd.log()` error. The only way to keep the website from changing is to not use it at all, and use the API instead.

Also, if `str.replace()` does not find any matches, it will not return an error.

Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: If definition is not found, then it goes to a different page. I'm sure you can detect something on that page for such detection. So if these tags don't exist and it doesn't find the tag on the definition not found page, then there is a problem.

File: sigma/plugins/utility/other/define.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        for i in args:
+            defURL += i
+            defURL += "%20"
+    else:
+        argsCopy = args.copy()
+        del(argsCopy[-1])
+        for i in argsCopy:
+            defURL += i
+            defURL += "%20"
+
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: This can just be defURL = r"https://www.merriam-webster.com/dictionary/" + args[0] and avoid the this entire if/for statement chain. Are there any words that have spaces?

Comment (johnmedina999) [reply sigma/plugins/utility/other/define.py]: There are (compound words such as `ice cream`), and that's the reason I have this loop.

File: sigma/plugins/utility/other/define.py (hunk 9 above)
Comment (johnmedina999) [reply sigma/plugins/utility/other/define.py]: Yes there are. Compound words (such as `ice cream`) are why I made the loop in the first place.

File: sigma/plugins/utility/other/define.py (hunk 10 above)
Comment (abraker95) [reply sigma/plugins/utility/other/define.py]: So if the number is for a different definition, wouldn't it be simpler to just pick the number randomly and drop the parameter?

Comment (johnmedina999) [reply sigma/plugins/utility/other/define.py]: Why would we want to do that though? It's always best to give the user control over what they're requesting. And this works fine.

File: sigma/plugins/utility/other/define.py (hunk 8 above)
Comment (abraker95) [reply sigma/plugins/utility/other/define.py]: I'm sure you can detect the page that comes up when the definition is not found. If it doesn't find anything associated with the definition or the lack of one, then assume the website has changed.

**Comment by johnmedina999** [comment]
Applied changes to `define.py` by creating a distinction between the definition being missing and the website encountering a 404 error, and the website changing. This way, we can know that the problem is, and we can fix it as soon as possible.

File: sigma/plugins/utility/other/define.py (hunk 8 above)
Comment (johnmedina999) [reply sigma/plugins/utility/other/define.py]: Done; is this alright?

File: sigma/plugins/utility/other/define.py (hunk 10 above)
Comment (abraker95) [reply sigma/plugins/utility/other/define.py]: Will it matter to the user if they don't know what will come up? Idk, it seems a but of unnecessary and slightly convoluted code for a not so important thing, imo

File: sigma/plugins/utility/other/define.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+            if(len(tag.next_sibling.strip()) == 0):
+                if hasPrinted:
+                    continue
+                for tag2  in soupObj.find_all(class_="sx-link sc"):
+
+                    defCopy = definition
+
+                    definition += tag2.text
+
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: Unnecessary whitespace

File: sigma/plugins/utility/other/define.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        return
+
+    definition.strip()
+    definition.replace(":"," \n— ")
+
+    if "The word you've entered" in definition:
+        embed = discord.Embed(title=":smile: That definition doesn't exist, silly!", color=0xFF8EE6)
+        await message.channel.send(None, embed=embed)
+        return
+    if "Sorry, the word you’re looking for" in definition:
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: I add spaces between ends and beginnings of multi-line conditional statements for readability

File: sigma/plugins/utility/other/define.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    
+    try:
+        rawDef = rawDef[rawDef.index("<!--JC: actual definitions-->")-10:]
+    except:
+        embed = discord.Embed(title=":exclamation: The information you entered does not correlate to a Dictionary Entry.", color=0xFF0000)
+        await message.channel.send(None, embed=embed)
+        return
+
+    rawDefArr = rawDef.split("<!--JC: actual definitions-->")
+
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: Unnecessary whitespace

File: sigma/plugins/utility/other/define.py (hunk 10 above)
Comment (johnmedina999) [reply sigma/plugins/utility/other/define.py]: By default, the first definition is given, which is usually enough to satisfy the necessity to use the command. However, some words can have radically different meanings, and this is where the option to retrieve a different meaning comes in handy. If we were to just give the user a random definition, it would not be very useful, because they might receive the definition they were looking for; and if they don't, they could always try again with a different number.

File: sigma/plugins/utility/other/define.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+    soupObj = BeautifulSoup(rawDefArr[defNum], "html.parser")
+
+    try:
+        for tag in soupObj.find_all(class_="intro-colon"):
+            definition += tag.text
+
+            if not definition:
+                raise IndexError("The website has changed, please fix.")
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: I am not sure if this will work. If this is true, then it wouldn't go to the for loop in the first place. If it goes to the for loop, then this must be false since definition is something in that case. Manually change the link to some other site to see if this works.

**Review by abraker95 — CHANGES_REQUESTED** [review]

File: sigma/plugins/utility/other/define.py (hunk 14 above)
Comment (johnmedina999) [reply sigma/plugins/utility/other/define.py]: This will work because nothing is being tested in order for the `for` loop to run.

Comment (abraker95) [reply sigma/plugins/utility/other/define.py]: The for loop will run if there are any "intro-colon" tags. If not, the definition wouldn't exist. Since you put `if not definition` inside the for loop, then when it's there, there must be a "intro-colon" tag, and therefore the definition exists due to the statement previous.

I highly recommend you test this code by changing the URL to some other site and report back the results.

File: sigma/plugins/utility/other/define.py (hunk 10 above)
Comment (abraker95) [reply sigma/plugins/utility/other/define.py]: Do the following instead to avoid this convoluted chain of if/for statements: 

```
tokens = args.split(" ")

 try: 
     defNum = int(tokens[-1])
     defURL += "%20".join(tokens[0:-1])
 except:
     defURL += "%20".join(tokens)
```

**Comment by johnmedina999** [comment]
Applied all requested changes.

**Comment by johnmedina999** [comment]
Applied all requested changes.

File: sigma/plugins/utility/other/define.py [hunk 15]
```diff
@@ -44,13 +44,14 @@
     rawDefArr = rawDef.split("<!--JC: actual definitions-->")
     soupObj = BeautifulSoup(rawDefArr[defNum], "html.parser")
 
-    try:
+    try:    
+        if '<span class="intro-colon">' in str(soupObj):
+            pass
+        else:
+            raise IndexError("The website has changed, please fix.")
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: Why not 
```
 if '<span class="intro-colon">' not in str(soupObj):
     raise IndexError("The website has changed, please fix.")
```

**Comment by abraker95** [comment]
Go up a bit, you missed one concerning the URL

File: sigma/plugins/utility/other/define.py (hunk 15 above)
Comment (johnmedina999) [reply sigma/plugins/utility/other/define.py]: I'm new to Python and didn't know you could do that, I will fix.

**Comment by johnmedina999** [comment]
Applied all requested changes.

**Review by johnmedina999 — COMMENTED** [review]
Version changes by itself, I don't change it.

File: sigma/plugins/utility/other/define.py [hunk 16]
```diff
@@ -0,0 +1,78 @@
+import discord
+import aiohttp
+from bs4 import BeautifulSoup
+
+async def define(cmd, message, args):
+
+    definition = ""
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: No need to predefine variables in python. They are defined the first time you use them.

File: sigma/plugins/utility/other/define.py [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    defNum = 1
+    hasPrinted = False
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py:9]: If you don't predefine, the default result of `if hasPrinted` is false, so this is redundent

File: sigma/plugins/utility/other/define.py [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+    soupObj = BeautifulSoup(rawDefArr[defNum], "html.parser")
+
+    try:    
+        if '<span class="intro-colon">' not in str(soupObj):
+            raise IndexError("The website has changed, please fix.")
+
+        for tag in soupObj.find_all(class_="intro-colon"):
+            definition += tag.text
+            if(len(tag.next_sibling.strip()) == 0):
+                if hasPrinted:
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: What does hasPrinted control?

Comment (johnmedina999) [reply sigma/plugins/utility/other/define.py]: I did not find a way to print any hyperlinked definitions in the order that they come in; therefore, I decided to print them all in one go whenever the loop comes to the first hyperlinked definition. The variable `hasPrinted` controls whether or not the hyperlinked definitions have all printed; this prevents them from printing again if the loop comes to another hyperlinked definition.

File: sigma/plugins/utility/other/define.py (hunk 17 above)
Comment (johnmedina999) [reply sigma/plugins/utility/other/define.py:9]: It's always better to be explicit about your variables, I find.

File: sigma/plugins/utility/other/define.py [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+                    continue
+                for tag2  in soupObj.find_all(class_="sx-link sc"):
+                    defCopy = definition
+                    definition += tag2.text
+                    if defCopy is definition:
+                        raise IndexError("The website has changed, please fix.")
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: This logic has the same mistake as the previous one you fixed; If it finds the "sx-link sc" tag, then definition will not be same after adding the contents inside the tag, and so `if defCopy is definition` is always false.

**Review by abraker95 — CHANGES_REQUESTED** [review]

File: sigma/plugins/utility/other/define.py (hunk 18 above)
Comment (abraker95) [reply sigma/plugins/utility/other/define.py]: Is it possible to put to avoid the nesting of the for statements and make two definition lists, one with the normal definitions and one with the hyperlinked definitions, and then join them together?

**Comment by johnmedina999** [comment]
Fixed bug, as requested.

File: sigma/plugins/utility/other/define.py (hunk 18 above)
Comment (johnmedina999) [reply sigma/plugins/utility/other/define.py]: That would simply remove the nesting; I can rewrite the section, if that's what you want.

File: sigma/plugins/utility/other/define.py [hunk 20]
```diff
@@ -35,11 +35,14 @@
...
     try:    
         if '<span class="intro-colon">' not in str(soupObj):
-            raise IndexError("The website has changed, please fix.")
+            raise IndexError('The website has changed, please fix. ("intro-colon")')
 
         for tag in soupObj.find_all(class_="intro-colon"):
             definition += tag.text
             if(len(tag.next_sibling.strip()) == 0):
+                if '<span class="sx-link sc">' not in str(soupObj):
+                    raise IndexError('The website has changed, please fix. ("sx-link sc")')
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: How about checking for this before the for loop?

Comment (johnmedina999) [reply sigma/plugins/utility/other/define.py]: That wouldn't work, because some word entries don't have hyperlinked definitions.

File: sigma/plugins/utility/other/define.py (hunk 18 above)
Comment (abraker95) [reply sigma/plugins/utility/other/define.py]: Preferably, yes. I try to avoid nesting as much as possible.

**Comment by abraker95** [comment]
Did you ever fix the definition texts looking like a text dump?

**Comment by johnmedina999** [comment]
No; for some reason, the `definition.replace(":"," \n— ")` on line 64 never runs.

**Comment by abraker95** [comment]
Try replacing ":" with something else. Does only that line never run or does it never find the colons to replace?

**Comment by johnmedina999** [comment]
It never runs, I tried it with a few other strings.

**Comment by johnmedina999** [comment]
Applied all requested changes.

File: sigma/plugins/utility/other/define.py [hunk 21]
```diff
@@ -35,12 +35,12 @@
 
     try:    
         if '<span class="intro-colon">' not in str(soupObj):
-            raise IndexError('The website has changed, please fix. ("intro-colon")')
+            raise ValueError('The website has changed, please fix. ("intro-colon")')
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: You can change this to just the generic Exception instead. Value nor Index errors make sense in this.

File: sigma/plugins/utility/other/define.py [hunk 22]
```diff
@@ -56,12 +56,16 @@
...
+        cmd.log('The website has changed; "sx-link sc" is not present.')
+        return
+    except ValueError:
+        cmd.log('The website has changed; "intro-colon" is not present.')
         return
 
     definition.strip()
-    definition.replace(":"," \n— ")
+    definition = definition.replace(":"," \n— ")
+    definition = definition.replace("\n", "", 1)
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: You can chain this to be
```
 definition = definition.replace(":"," \n— ").replace("\n", "", 1)
```

File: sigma/plugins/utility/other/define.py [hunk 23]
```diff
@@ -56,12 +56,16 @@
             else:
                 definition += tag.next_sibling
                 definition += " "
-    except:
-        cmd.log('The website has changed; either "intro-colon" or "sx-link sc" are not present.')
+    except IndexError:
+        cmd.log('The website has changed; "sx-link sc" is not present.')
+        return
+    except ValueError:
```
Comment (abraker95) [inline sigma/plugins/utility/other/define.py]: The string you specify as a parameter of the exceptions such as ValueError(...) or Exception(...) or any other can be retrieved by casting the exception into a string. So just keep one, change to Exception and print whatever it is.
```
except Exception as e
      cmd.log(e)
      return
```

**Review by abraker95 — CHANGES_REQUESTED** [review]

**Comment by johnmedina999** [comment]
Applied all requested changes.

**Review by abraker95 — APPROVED** [review]
Looks good! Will merge in the changes soon.
